Zend/zend_compile.c

```diff
@@ -7365,6 +7464,50 @@ void zend_compile_class_decl(znode *result, zend_ast *ast, zend_bool toplevel) /
 			zend_resolve_const_class_name_reference(extends_ast, "class name");
 	}
 
+	if (enum_primitive_type_ast != NULL) {
+		ZEND_ASSERT(ce->ce_flags & ZEND_ACC_ENUM);
+		zend_type type = zend_compile_typename(enum_primitive_type_ast, 0, 0);
+		if (
+			type.ptr != NULL
+			|| (
+				type.type_mask != MAY_BE_LONG
+				&& type.type_mask != MAY_BE_STRING
+			)
+		) {
+			zend_error_noreturn(E_ERROR, "Enum primitive type must be int or string");
+		}
+		if (type.type_mask == MAY_BE_LONG) {
+			ce->enum_primitive_type = IS_LONG;
+		} else if (type.type_mask == MAY_BE_STRING) {
+			ce->enum_primitive_type = IS_STRING;
+		} else {
+			ZEND_ASSERT(0);
+		}
+		zend_type_release(type, 0);
+
+		ce->enum_case_table = emalloc(sizeof(HashTable));
+		zend_hash_init(ce->enum_case_table, 0, NULL, ZVAL_PTR_DTOR, 0);
+	}
+
+	if (ce->ce_flags & ZEND_ACC_ENUM) {
+		if (implements_ast == NULL) {
+			implements_ast = zend_ast_create_list(0, ZEND_AST_NAME_LIST);
+			decl->child[1] = implements_ast;
+		}
+		
+		zend_string *unit_enum_string = zend_string_init("UnitEnum", strlen("UnitEnum"), 0);
+		zend_ast *unit_enum_ast = zend_ast_create_zval_from_str(unit_enum_string);
+		unit_enum_ast->attr = ZEND_NAME_FQ;
+		zend_ast_list_add(implements_ast, unit_enum_ast);
+
+		if (ce->enum_primitive_type != IS_UNDEF) {
+			zend_string *scalar_enum_string = zend_string_init("ScalarEnum", strlen("ScalarEnum"), 0);
```

Comment:
Relatedly, this could be an if-else to implement one interface or the other if ScalarEnum extended the interface UnitEnum

Reply:
Implementing both explicitly seems fine to me.